File: frontend/src/lib/utils/chat/eviction.ts
/**
 * Eviction policy: what the client does once a Remove commit naming this device has been applied.
 *
 * The rule is that the REMOVE COMMIT IS AUTHORITATIVE. It is a signed, ordered statement by a
 * member with the right to make it, and it is the same statement every other member applies - so
 * there is nothing to confirm, nothing to retry and nothing to repair. The conversation is retired
 * exactly as a peer-side deletion retires it, and `requestReAdd` is reserved for what it was
 * written for: a group this device believes it is IN but cannot use (an epoch fork, a lost
 * Welcome). Eviction is not a broken state, it is a correct one we are not part of.
 *
 * The fact is read from OpenMLS (`isGroupActive`), never mirrored into a flag of our own: the group
 * state is already durable, and a second copy can only ever be wrong in the direction that matters
 * - saying we are still a member of a group we were removed from.
 *
 * ## The fact is learnt in five places and RECORDED in one
 *
 * A client can find out it is no longer a member six different ways, at five sites: a Remove commit
 * it applies, the exclusion its author announces, an inbound frame refused as `evicted`, a
 * membership check made before a send or a selection, a send the server refuses, and the delivery
 * service answering `NotAGroupMemberError`. Each of those is a different EVIDENCE, and that is the
 * only thing they legitimately differ in. What is OWED once the fact is known is the same every
 * time, and it was written out by hand at every site until it stopped agreeing with itself:
 *
 * | Site | Retired the row | Posted the notice |
 * | --- | --- | --- |
 * | a Remove commit applied, and an inbound frame refused as evicted | yes | no |
 * | the exclusion announced by its author (`memberRemoved`) | yes | yes |
 * | a send refused, before or by the server (the outbox) | yes | no |
 * | a membership check on selection, and the server's own refusal | **no** | yes |
 *
 * The last row is the defect, and it is the one the user meets: `verifyCurrentUserMembership`
 * learnt the eviction from the authoritative local state, posted a notice, cached "not a member"
 * for thirty seconds - and wrote NOTHING down. A reload brought the conversation back as `active`,
 * the roster request went back to the members-only endpoint, and the fact had to be re-learnt on
 * the next selection, for ever. {@link recordEviction} is the single disposition; the evidence
 * travels as data, and it is the only thing a caller decides.
 *
 * DROPPING THE MLS STATE IS NOT PART OF IT, and deliberately so. That is its own operation with its
 * own single implementation (`dropGroupState`), and an evicted group's state is not worthless: it
 * is what still opens the frames of the epochs this device WAS a member for, and it is what makes a
 * later frame classify as `evicted` at all rather than falling through to the out-of-sync policy.
 * The row is what refuses the send, the roster, the re-add and the welcome_request - so the two
 * axes are separate, and `memberRemoved` is simply a site that legitimately does both.
 */

import type { IMlsService } from '$lib/mls-client/IMlsService';
import { m } from '$lib/paraglide/messages';
import type { AddMessageToChatOptions, Conversation } from '$lib/types';
import { findConversationKeyByGroupId, retireConversation } from '$lib/utils/chat/conversations';

/**
 * How this device found out, and the ONLY thing the five learning sites legitimately differ in.
 *
 * It is not decoration. Two of these six mean the fact was learnt by FAILING - the Remove commit
 * that stated it never reached this device, so a message was written, encrypted and refused to
 * find out what a frame already knew. That is the one thing a reader of the log needs to be able
 * to tell apart, and until this table existed two sites said so in prose and four said nothing.
 */
export type EvictionEvidence =
  /** A Remove commit naming this device merged, and the membership read right after it says so. */
  | 'remove-commit'
  /** The member who made the removal announced it (`memberRemoved` naming us). */
  | 'system-event'
  /** An inbound frame for the group was refused as `evicted` - in flight, or routed by a stale registry. */
  | 'inbound-frame'
  /** A membership check - before a send, or on selecting the conversation - answered `false`. */
  | 'membership-check'
  /** A send was refused as `evicted` AFTER a membership check answered yes. A miss. */
  | 'outbound-refusal'
  /** The delivery service refused this device as a non-member. A miss. */
  | 'server-refusal';

/**
 * What each evidence says in the log, and whether reaching it means the fact was learnt by failing.
 *
 * The shape of the line is the classification: a `missed` evidence never produces
 * `[EVICT] Removed from ...`, so no rule written for the healthy path can swallow it, and the
 * campaign's watcher leaves it unexplained - which is what it is.
 */
const EVICTION_EVIDENCE: Record<EvictionEvidence, { what: string; missed: boolean }> = {
  'remove-commit': { what: 'by a Remove commit', missed: false },
  'system-event': { what: 'by an exclusion its author announced', missed: false },
  'inbound-frame': {
    what: 'by a Remove commit a frame for the group arrived after',
    missed: false,
  },
  'membership-check': { what: 'by a Remove commit this device had already applied', missed: false },
  'outbound-refusal': { what: 'a send was refused as evicted', missed: true },
  'server-refusal': {
    what: 'the delivery service refused this device as a non-member',
    missed: true,
  },
};

/** What {@link recordEviction} needs. `groupId` and not a key: four of the five sites hold nothing else. */
export interface RecordEvictionDeps {
  conversations: Map<string, Conversation>;
  /** The MLS group id. The row is found by it. */
  groupId: string;
  /** How this device found out - see {@link EvictionEvidence}. */
  evidence: EvictionEvidence;
  log: (message: string) => void;
  saveConversation?: (key: string) => Promise<void>;
  /**
   * Posts the one-time removal notice.
   *
   * Optional because a caller may have no chat surface at all, never because the notice is. A site
   * that omits it leaves the user with the banner and no line in the thread, which is exactly the
   * disagreement this function exists to end - so every site that can supply it, does.
   */
  addMessageToChat?: (
    senderId: string,
    content: string,
    contactName: string,
    options?: AddMessageToChatOptions
  ) => Promise<void>;
}

/**
 * Records that this device is no longer a member of a group - the ONE disposition, whatever the
 * evidence.
 *
 * Three things, in this order, and all three are owed every time:
 *
 *   1. **the row is retired**, which is the durable statement. `lifecycle: 'removed'` is read by
 *      the send gate, the roster loader, the frame buffer, the re-add guards and the UI banner, so
 *      recording the eviction anywhere else - or only in memory - leaves every one of them wrong.
 *   2. **the notice is posted once**, so the user is told rather than left with a conversation that
 *      silently refuses to send. Deduplicated on the message itself, because the same eviction is
 *      legitimately learnt twice (a commit and then the announcement behind it).
 *   3. **one line naming the evidence**, and accusing when the evidence is a miss.
 *
 * Idempotent: a replayed commit, or a second evidence for the same removal, changes nothing and
 * says so. Returns whether THIS call performed the transition.
 */
export async function recordEviction(deps: RecordEvictionDeps): Promise<boolean> {
  const { conversations, groupId, evidence, log, saveConversation, addMessageToChat } = deps;
  const short = groupId.slice(0, 8);
  const { what, missed } = EVICTION_EVIDENCE[evidence];

  const key = findConversationKeyByGroupId(conversations, groupId);
  if (!key) {
    // Not silent, and not an error either: a distribution group and a group whose row was already
    // purged both reach here. Nothing is recorded about them, so there is nothing to retire.
    log(`[EVICT] ${short}… - removed (${what}), but no conversation row carries this group`);
    return false;
  }

  // Read BEFORE the retire, which rewrites the row: the notice is deduplicated on the thread's own
  // contents, the only place the first evidence left a trace the second can see.
  const notice = m.chat_system_removed_from_group();
  const alreadyNoticed =
    conversations.get(key)?.messages.some((msg) => msg.isSystem && msg.content === notice) ?? false;

  const retired = await retireConversation({ conversations, key, groupId, saveConversation });

  if (!alreadyNoticed) {
    // Best-effort, and after the retire: the banner is the durable half and must not be held up by
    // a message write. A swallowed branch logs.
    await addMessageToChat?.('system', notice, key, { isSystem: true }).catch((e: unknown) =>
      log(
        `[EVICT] ${short}… - removal notice not posted: ${e instanceof Error ? e.message : String(e)}`
      )
    );
  }

  if (missed) {
    // NEVER LEARN BY FAILING WHAT A FACT COULD HAVE TOLD YOU. The Remove commit states this before
    // anything is sent; reaching here means it never arrived, and the line has to accuse so the
    // path stays a signal rather than becoming the way eviction is normally discovered.
    log(
      `[EVICT] ${short}… - ${what}, so this device learnt its removal by FAILING: the Remove commit never reached it`
    );
  } else {
    log(
      retired
        ? `[EVICT] Removed from ${short}… ${what} - conversation retired`
        : `[EVICT] Removed from ${short}… - already retired, nothing to do`
    );
  }
  return retired;
}

/** Everything {@link retireIfEvicted} needs, kept to the narrowest set both call sites can supply. */
export interface EvictionCheckDeps extends Omit<RecordEvictionDeps, 'groupId' | 'evidence'> {
  mlsService: Pick<IMlsService, 'isGroupActive'>;
  groupId: string;
  /**
   * How this device came to be asking.
   *
   * Narrowed to the three evidences a LOCAL READ can establish: the two misses already carry the
   * server's own word, so a caller holding one has nothing left to ask OpenMLS, and `system-event`
   * is the removal's author stating it. Passing one of those here would re-read a fact already in
   * hand and answer `false` on a group this device does not hold.
   */
  evidence: Extract<EvictionEvidence, 'remove-commit' | 'inbound-frame' | 'membership-check'>;
}

/**
 * Reads this device's membership of `groupId` from OpenMLS - the authoritative fact.
 *
 * THREE ANSWERS, NOT TWO, and the third is the point. `true` we are a member, `false` a Remove
 * commit named us, `null` the local state could not say - a group this device does not hold throws
 * here, and collapsing that into `false` would retire conversations it had merely not loaded yet.
 * Every caller must decide what to do about `null` for itself; none of them may treat it as a "no".
 *
 * This is the fact `verifyCurrentUserMembership` consults before it asks the delivery service, and
 * the reason it can: the answer is local, durable and free, and the server's own membership
 * endpoint is members-only - so on the one question that matters it is certain to refuse.
 *
 * The swallowed branch logs, and names its caller through `context`: it is the branch that would
 * hide an eviction, so a run must be able to see which decision it was about to inform.
 */
export async function readLocalMembership(deps: {
  mlsService: Pick<IMlsService, 'isGroupActive'>;
  groupId: string;
  /** What was about to be decided, spliced into the log line ("after a commit"). */
  context: string;
  log: (message: string) => void;
}): Promise<boolean | null> {
  const { mlsService, groupId, context, log } = deps;
  try {
    return await mlsService.isGroupActive(groupId);
  } catch (e) {
    log(`[EVICT] Membership of ${groupId.slice(0, 8)}… could not be read ${context}: ${String(e)}`);
    return null;
  }
}

/**
 * Retires the conversation if this device is no longer a member of the group.
 *
 * Returns true when this call performed the transition, false when the device is still a member,
 * the conversation was already retired, or membership could not be read. Idempotent: the retire
 * itself is a no-op on a conversation already in `removed`, so a replayed commit costs one query.
 *
 * A failure to READ membership is deliberately not treated as an eviction. The two are opposite
 * facts and only one of them retires a conversation; an unheld group throws here, and answering
 * "evicted" to that would retire conversations this device simply had not loaded yet.
 */
export async function retireIfEvicted(deps: EvictionCheckDeps): Promise<boolean> {
  const { mlsService, groupId, evidence, log, ...rest } = deps;
  // `null` is NOT an eviction: a membership query is not the point of the path this runs on, and the
  // send-path backstop in the outbox is what would find a missed eviction, one refused message
  // later. `readLocalMembership` owns the log for that branch.
  const active = await readLocalMembership({ mlsService, groupId, context: 'after a commit', log });
  if (active !== false) return false;

  return recordEviction({ ...rest, groupId, evidence, log });
}

/**
 * Whether the row itself already records that this device is OUT of the group - so the delivery
 * service must not be asked anything that only a member may ask.
 *
 * WHY A SECOND READER OF THE SAME FACT. `readLocalMembership` answers the question for a caller
 * that can afford an async round-trip into OpenMLS and that has something to do with all three of
 * its answers. The roster loader has neither: it runs on every conversation SELECTION, it wants one
 * cheap yes/no, and the only thing it can do with "no" is show no roster. `lifecycle === 'removed'`
 * is that yes/no, it is durable, it is already in the row the caller is holding, and it is written
 * in exactly one place (`retireConversation`) from the Remove commit that is authoritative.
 *
 * IT EXISTS BECAUSE THE GUARD WAS PUT ON ONE OF TWO DOORS. On 2026-08-23 the membership check
 * stopped asking `GET /api/mls/groups/:id/members` on a device holding a Remove commit, because
 * that endpoint is members-only BY DESIGN and could only refuse. `loadGroupMembers` reaches the
 * same endpoint, is fired by the same two selection paths one line above that check, and had no
 * guard of any kind - so a removed device selecting its retired conversation still logged the same
 * 403, and GRP-3 still recorded it on 2026-08-24. Fixing a call site is not fixing a seam.
 *
 * IT COVERS A DEPARTURE THIS DEVICE CHOSE, and only because the departure states the fact before
 * it acts. A leave and a delete are not learnt from a commit - the device decides - and they used
 * to record that decision LAST, after the server call and the WASM forget, which left a window in
 * which the row still read as live while the membership behind it was already gone. `$effect`s over
 * the conversations map fire inside that window, so the same 403 came back a third time (GRP-6,
 * 2026-08-24). `exitGroupAndCleanup` now retires first and purges last; this predicate did not have
 * to widen, and a third call-site guard was not what was missing.
 *
 * A retired conversation has no roster this device is entitled to know, so `false` here is the
 * ANSWER and not a fallback: the absence is what the UI should render.
 *
 * `undefined` - a group id naming no row - is deliberately not "lost". Nothing is recorded about
 * it, and suppressing a request on the strength of a missing row would hide a real lookup bug.
 */
export function membershipIsDurablyLost(
  convo: Pick<Conversation, 'lifecycle'> | undefined
): boolean {
  return convo?.lifecycle === 'removed';
}

File: frontend/mls-core/tests/eviction.rs
//! Being removed from a group, from the removed device's own point of view.
//!
//! The Remove commit NAMES the device it evicts. Applying it is therefore the instant that device
//! could know it is out - and until this was written, that instant produced nothing: the merge
//! answered `Ok(None)`, exactly as any other structural commit does, and the only thing left that
//! could tell the two apart was the next send being refused.
//!
//! That is learning by failing what a fact already stated, and it cost more than a diagnosis. The
//! refusal crossed the FFI boundary as `MlsError::OpenMls("Encrypt error: ... UseAfterEviction")`,
//! indistinguishable from a transient encrypt failure, so the frontend outbox retried it up its
//! whole backoff ladder against a group that would refuse every single attempt.
//!
//! FOUR PATHS REACH AN EVICTED DEVICE, and all four are pinned here because the fix is only a fix
//! if all four agree:
//!
//!   - the COMMIT, which is where the fact becomes knowable and free (`is_group_active`);
//!   - the SEND, for a device that never received that commit - typed, not a sentence
//!     (`MlsError::Evicted`), so a permanent failure cannot be read as a transient one;
//!   - the RECEIVE, for a frame still routed to a group we are out of. Unclassified it read as
//!     "out of sync", which asked to be re-added to a group we were deliberately removed from and
//!     then learnt from a 403 what the frame itself already proved;
//!   - the RECEIVE OF A LATER EPOCH, which is the SAME frame arriving after the group has committed
//!     again - and the commonest shape of it, because a group does not stop moving when it loses a
//!     member. The epoch-gap fast-fail sits before the decryption that would reveal the eviction,
//!     so this path answered "gap" and reached the same out-of-sync policy by a different road.
use mls_core::{DecryptErrorKind, MlsError, MlsManager};

fn make_device(user_id: &str, device_id: &str) -> MlsManager {
    MlsManager::load_or_create(user_id, device_id, None)
        .unwrap_or_else(|e| panic!("device '{user_id}:{device_id}': {e}"))
}

/// alice creates `gid` with bob in it, and both devices are left at the same epoch.
fn pair(gid: &str) -> (MlsManager, MlsManager) {
    let mut alice = make_device("alice", "dev1");
    alice.create_group(gid.to_string()).expect("create_group");

    let mut bob = make_device("bob", "dev1");
    let kp = bob.generate_key_package().expect("kp");
    let (_commit, welcome, _added, _skipped) =
        alice.add_members_bulk(gid, &[kp.as_slice()]).expect("add");
    alice.merge_pending_commit_for(gid).expect("merge add");
    let tree = alice.export_ratchet_tree_for(gid).expect("tree");
    bob.process_welcome(welcome.as_deref().unwrap(), Some(&tree))
        .expect("bob joins");
    (alice, bob)
}

/// alice removes bob and returns the commit, merged on alice's side.
fn evict_bob(alice: &mut MlsManager, gid: &str) -> Vec<u8> {
    let commit = alice
        .remove_members_for_devices(gid, &["bob:dev1"])
        .expect("remove bob");
    alice.merge_pending_commit_for(gid).expect("merge remove");
    commit
}

#[test]
fn the_removed_device_reads_its_own_eviction_off_the_commit() {
    let gid = "g-evict-read";
    let (mut alice, mut bob) = pair(gid);

    assert!(
        bob.is_group_active(gid).expect("bob holds the group"),
        "a member is active before anything removes it"
    );

    let commit = evict_bob(&mut alice, gid);
    bob.process_incoming_message(gid, &commit)
        .expect("bob applies the commit that removes him");

    assert!(
        !bob.is_group_active(gid).expect("bob still HOLDS the state"),
        "the Remove commit named this device: applying it is when the eviction becomes knowable, \
         and it must be knowable without attempting a send"
    );
}

#[test]
fn the_remover_stays_active_in_the_group_it_shrank() {
    // The obvious half, and worth pinning: `is_group_active` must answer membership, not "did a
    // Remove commit happen here". A predicate that went false for everyone on any removal would
    // retire the remover's own conversation.
    let gid = "g-evict-remover";
    let (mut alice, _bob) = pair(gid);
    evict_bob(&mut alice, gid);

    assert!(
        alice.is_group_active(gid).expect("alice holds the group"),
        "removing another member does not remove yourself"
    );
}

#[test]
fn holding_a_group_and_being_in_it_are_different_questions() {
    // The removed device still HOLDS local state for the group - which is precisely why the
    // frontend's `getLocalGroups().includes(id)` stayed true after an eviction and let the outbox
    // conclude the group was sendable.
    let gid = "g-evict-held";
    let (mut alice, mut bob) = pair(gid);
    let commit = evict_bob(&mut alice, gid);
    bob.process_incoming_message(gid, &commit).expect("applies");

    assert!(
        bob.get_known_groups().contains(&gid.to_string()),
        "the state is still held after an eviction - it is simply no longer usable"
    );
    assert!(!bob.is_group_active(gid).expect("held"), "and not a member");
}

#[test]
fn a_group_never_joined_is_an_error_not_an_eviction() {
    // Never-joined and removed-from are opposite facts and only one of them retires a conversation.
    // Answering `false` for an unheld group would retire every conversation not yet loaded.
    let alice = make_device("alice", "dev1");
    assert!(
        matches!(
            alice.is_group_active("g-never-created"),
            Err(MlsError::GroupNotFound(_))
        ),
        "an unheld group is an error, never `not a member`"
    );
}

#[test]
fn a_send_after_eviction_is_typed_evicted_not_a_generic_encrypt_error() {
    // The backstop, for a device that was offline across the whole removal and sends before it
    // drains its inbox: it never applied the commit, so `is_group_active` cannot help and OpenMLS
    // refuses the send. That refusal must arrive as a TYPE. Carried as `OpenMls(String)` it was
    // read as transient - the one classification that turns a permanent failure into a retry loop.
    let gid = "g-evict-send";
    let (mut alice, mut bob) = pair(gid);
    let commit = evict_bob(&mut alice, gid);
    bob.process_incoming_message(gid, &commit).expect("applies");

    let err = bob
        .send_message(gid, b"nobody will ever read this")
        .expect_err("an evicted device cannot encrypt for the group");

    assert!(
        matches!(err, MlsError::Evicted(ref g) if g == gid),
        "expected MlsError::Evicted({gid}), got {err:?}"
    );
    // The token is the contract across both FFI boundaries: `classifyOutgoingSendError` reads it,
    // and the outbox fails the entry permanently on it.
    assert!(
        err.to_string().starts_with("EVICTED: "),
        "the discriminator must survive the string-only WASM/Tauri boundary, got {err}"
    );
}

#[test]
fn a_healthy_member_sends_normally() {
    // Guards the arm above against the other kind of mistake: classifying too much as an eviction.
    let gid = "g-evict-control";
    let (_alice, mut bob) = pair(gid);
    assert!(
        bob.send_message(gid, b"hello").is_ok(),
        "a member still in the group encrypts as before"
    );
}

#[test]
fn a_frame_arriving_after_eviction_is_typed_evicted_not_an_out_of_sync() {
    // THE THIRD PATH, and the one that cost the most. A frame still reaches the removed device -
    // in flight when the commit landed, or routed by the server registry the removal cleans
    // best-effort - and `process_message` refuses it because the group is inactive.
    //
    // Unclassified, that refusal was a bare `Process error:`, which the frontend reads as
    // `SenderRatchetGap`/`unknown` and answers with its out-of-sync policy: `requestReAdd`, asking
    // the server to undo a moderation action, followed by a request for the group's commits that
    // can only ever return 403 - learning from a refusal what this very frame already proved.
    let gid = "g-evict-recv";
    let (mut alice, mut bob) = pair(gid);
    let commit = evict_bob(&mut alice, gid);
    bob.process_incoming_message(gid, &commit).expect("applies");

    // alice keeps talking to the group bob is no longer in.
    let frame = alice
        .send_message(gid, b"members only")
        .expect("alice sends");
    let err = bob
        .process_incoming_message(gid, &frame)
        .expect_err("bob cannot read a group he is not in");

    assert!(
        matches!(err, MlsError::Evicted(ref g) if g == gid),
        "expected MlsError::Evicted({gid}), got {err:?}"
    );
    assert_eq!(
        err.decrypt_kind(),
        DecryptErrorKind::Evicted,
        "native and web classify through decrypt_kind and classifyIncomingDecryptError \
         respectively; the two must not be able to disagree about this one"
    );
}

#[test]
fn eviction_is_classified_before_the_generic_process_error_arm() {
    // Order matters here exactly as it does for the four arms already documented in
    // `decrypt_kind`. A frame refused for eviction IS a process error, so a generic arm reached
    // first would classify it `SenderRatchetGap` - the retryable kind - and native would write a
    // row into `pending_mls_messages` for a frame that can never decrypt, on every frame the group
    // still routes. This pins the precedence rather than trusting the arm order to survive edits.
    let raw = MlsError::OpenMls("Process error: GroupStateError(UseAfterEviction)".into());
    assert_eq!(
        raw.decrypt_kind(),
        DecryptErrorKind::SenderRatchetGap,
        "the raw wording is deliberately NOT what carries the meaning - the variant is"
    );
    assert_eq!(
        MlsError::Evicted("g".into()).decrypt_kind(),
        DecryptErrorKind::Evicted
    );
}

#[test]
fn a_later_epoch_frame_after_eviction_is_evicted_not_an_epoch_gap() {
    // THE FOURTH PATH, and the one the other three hid. The receive arm above is reached only for a
    // frame at OUR epoch: the epoch-gap fast-fail returns first for anything ahead of it, before
    // any decryption that could reveal the eviction. And a group keeps committing after it drops a
    // member, so being ahead is the NORMAL state of every frame that still reaches the removed
    // device - the classified path was the exception, not the rule.
    //
    // "Gap" then read as out-of-sync, which is `requestReAdd` against a deliberate removal plus a
    // commit request that can only 403. `is_active()` is local state and already false, so nothing
    // had to be attempted to know better.
    let gid = "g-evict-later-epoch";
    let (mut alice, mut bob) = pair(gid);
    let commit = evict_bob(&mut alice, gid);
    bob.process_incoming_message(gid, &commit).expect("applies");

    // The group moves on without bob: one more commit, so alice is now an epoch ahead of him.
    let carol = make_device("carol", "dev1");
    let kp = carol.generate_key_package().expect("kp");
    alice
        .add_members_bulk(gid, &[kp.as_slice()])
        .expect("add carol");
    alice
        .merge_pending_commit_for(gid)
        .expect("merge add carol");

    let frame = alice
        .send_message(gid, b"members only, one epoch later")
        .expect("alice sends");
    let err = bob
        .process_incoming_message(gid, &frame)
        .expect_err("bob cannot read a group he is not in");

    assert!(
        matches!(err, MlsError::Evicted(ref g) if g == gid),
        "a frame from a LATER epoch is still an eviction, not a gap to recover from: {err:?}"
    );
    assert_eq!(
        err.decrypt_kind(),
        DecryptErrorKind::Evicted,
        "an epoch gap is the retryable kind; charging this frame to it asks the server to undo a          moderation action, once per frame the group still routes"
    );
}

#[test]
fn a_later_epoch_frame_in_a_group_we_are_still_in_is_still_a_gap() {
    // The other half, and the one that keeps the hoist honest: eviction is now decided before the
    // gap, so a genuine gap must survive that. A member who simply missed a commit is exactly the
    // case the fast-fail exists for, and misreading it as an eviction would retire a live
    // conversation on a recoverable condition.
    let gid = "g-gap-still-member";
    let (mut alice, mut bob) = pair(gid);

    // alice commits without bob hearing about it, then talks at the new epoch.
    let carol = make_device("carol", "dev1");
    let kp = carol.generate_key_package().expect("kp");
    alice
        .add_members_bulk(gid, &[kp.as_slice()])
        .expect("add carol");
    alice
        .merge_pending_commit_for(gid)
        .expect("merge add carol");
    let frame = alice
        .send_message(gid, b"bob missed a commit")
        .expect("send");

    let err = bob
        .process_incoming_message(gid, &frame)
        .expect_err("bob is an epoch behind");

    assert!(
        bob.is_group_active(gid).expect("bob still holds it"),
        "missing a commit does not remove anybody"
    );
    assert!(
        !matches!(err, MlsError::Evicted(_)),
        "a member who missed a commit has a gap to recover, not an eviction: {err:?}"
    );
    assert_eq!(
        err.decrypt_kind(),
        DecryptErrorKind::SenderRatchetGap,
        "the gap is the retryable kind, and must stay retryable"
    );
}
